### PR TITLE
Add i.MX8M HW support

### PR DESCRIPTION
**Pull request by dbaluta**
This is initial skeleton support. i.MX8M is very different from i.MX8. Biggest difference that impacts SOF imx code is the way resources are managed:
* i.MX8M - Linux manages the resources
* i.MX8 - a special core manages the resources and runs a special firmware named SCFW

Cc: @paulstelian97 @TE-N-ShengjiuWang

### CODE REVIEW

File: sound/soc/sof/imx/imx8m.c [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+	struct imx8m_priv *priv;
+	struct resource res;
+	u32 base, size;
+	int ret = 0;
+
+	priv = devm_kzalloc(&pdev->dev, sizeof(*priv), GFP_KERNEL);
+	if (!priv)
+		return -ENOMEM;
+
+	sdev->private = priv;
```
Comment (paulstelian97) [inline sound/soc/sof/imx/imx8m.c:136]: Memory leak with the return on line 144? Or does devm clean up on probe failure?

Comment (dbaluta) [reply sound/soc/sof/imx/imx8m.c:136]: That's the role of managed memory allocation, it will cleanup the memory on error or on remove.

Comment (paulstelian97) [reply sound/soc/sof/imx/imx8m.c:136]: Alright and -EPROBE_DEFER does also fit the error case, thanks.

Comment (dbaluta) [reply sound/soc/sof/imx/imx8m.c:136]: @paulstelian97 devm_kzalloc documentation says that memory will be freed at device detach. I assume that the core creates a device, does attach and then calls probe. If probe fails then it does detach which in turn cleans up the memory. 

This happens on error path.

Another, case where memory is freed is when device is intentionally detached which happens when remove function is called.

**Comment by dbaluta** [comment]
@paulstelian97 weren't you supposed to stay at home and learn for your exams? :D

**Comment by paulstelian97** [comment]
> @paulstelian97 weren't you supposed to stay at home and learn for your exams? :D

The one I had today actually went really well, thanks :) Guess I can learn for the others

File: sound/soc/sof/imx/imx8m.c [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+static void imx8m_dsp_handle_reply(struct imx_dsp_ipc *ipc)
+{
+	struct imx8m_priv *priv = imx_dsp_get_data(ipc);
+	unsigned long flags;
+
+	spin_lock_irqsave(&priv->sdev->ipc_lock, flags);
+	imx8m_get_reply(priv->sdev);
+	snd_sof_ipc_reply(priv->sdev, 0);
+	spin_unlock_irqrestore(&priv->sdev->ipc_lock, flags);
```
Comment (plbossart) [inline sound/soc/sof/imx/imx8m.c:85]: out of curiosity, do you handle this lock in interrupt context?

Comment (dbaluta) [reply sound/soc/sof/imx/imx8m.c:85]: @plbossart yes, this is a stacktrace:

```
[   11.247874] Hardware name: FSL i.MX8MP EVK (DT)
[   11.252402] Call trace:
[   11.254853]  dump_backtrace+0x0/0x158
[   11.258514]  show_stack+0x14/0x20
[   11.261830]  dump_stack+0xbc/0x100
[   11.265234]  imx8_dsp_handle_reply+0x48/0x178 [snd_sof_imx8m]
[   11.270980]  imx_dsp_handle_rx+0x50/0x60
[   11.274903]  mbox_chan_received_data+0x18/0x28
[   11.279346]  imx_mu_isr+0x1b4/0x248
[   11.282836]  __handle_irq_event_percpu+0x70/0x198
[   11.287540]  handle_irq_event_percpu+0x34/0x88
[   11.291983]  handle_irq_event+0x44/0xc8
[   11.295818]  handle_fasteoi_irq+0xac/0x158
[   11.299914]  generic_handle_irq+0x24/0x38
[   11.303923]  __handle_domain_irq+0x5c/0xb8
[   11.308018]  gic_handle_irq+0x5c/0x150
[   11.311765]  el1_irq+0xb8/0x180
[   11.314907]  cpuidle_enter_state+0x138/0x340
[   11.319176]  cpuidle_enter+0x38/0x50
[   11.322751]  call_cpuidle+0x1c/0x48
[   11.326238]  do_idle+0x218/0x2b8
[   11.329465]  cpu_startup_entry+0x24/0x40
[   11.333388]  rest_init+0xd4/0xe0
[   11.336616]  arch_call_rest_init+0xc/0x14
[   11.340625]  start_kernel+0x42c/0x458
```

File: sound/soc/sof/imx/imx8m.c [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+		dev_err(&pdev->dev, "failed to get memory region node\n");
+		ret = -ENODEV;
+		goto exit_pdev_unregister;
+	}
+
+	ret = of_address_to_resource(res_node, 0, &res);
+	if (ret) {
+		dev_err(&pdev->dev, "failed to get reserved region address\n");
+		goto exit_pdev_unregister;
+	}
```
Comment (lyakh) [inline sound/soc/sof/imx/imx8m.c:188]: Looks like you could move calls to `platform_get_resource()`, `of_parse_phandle()`, `of_address_to_resource()` to the beginning of the function and simplify the error handling - if any of them fails, you'd just return the error?

File: sound/soc/sof/imx/imx8m.c [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+				struct snd_pcm_substream *substream,
+				const struct sof_ipc_pcm_params_reply *reply)
+{
+	return 0;
+}
+
+static struct snd_soc_dai_driver imx8m_dai[] = {
+{
+	.name = "sai-port",
+},
```
Comment (lyakh) [inline sound/soc/sof/imx/imx8m.c:243]: I'd add one more level of indentation, or you could actually just declare 1 instance of `struct snd_soc_dai_driver` without making it an array.

Comment (dbaluta) [reply sound/soc/sof/imx/imx8m.c:243]: I prefer to leave it as an array, as we can add more dais in the future. Also, by one more level of indentation you mean:

```
static struct snd_soc_dai_driver imx8m_dai[] = {
       {
	     .name = "sai-port",
       },
};
```

I can actually do that. For now I was just keeping the same pattern as in sound/soc/sof/intel ..

File: sound/soc/sof/imx/imx8m.c (hunk 3 above)
Comment (dbaluta) [reply sound/soc/sof/imx/imx8m.c:188]: @lyakh yes, it makes sense. will fix in v2. Also I think I can replace platform_get_resource + devm_ioremap with devm_ioremap_resource

File: sound/soc/sof/imx/imx8m.c (hunk 4 above)
Comment (lyakh) [reply sound/soc/sof/imx/imx8m.c:243]: I see, well, up to you really, indeed hda-dai.c and others use that indentation style...

**Pull request merged**